### ISSUE TITLE
XSRF token is not generated in the frontend

### ISSUE THREAD

**Issue by michael-e** [symphonycms/symphony-2#2173]
In the frontend (i.e. without a visitor being logged in to Symphony), XSRF tokens are not generated. The backend uses `$this->Page->Form->prependChild(XSRF::formToken());` on every form, which will in turn call `self::getToken()`—so a token gets generated. But in the frontend, this doesn't happen (since we don't want to automatically inject any token markup). My first idea to solve this is to add:

```
XSRF::getToken();
```

in line 422 of `class.frontendpage.php` (immediately before adding the cookie values).

However, I am not sure if this is the correct approach.

**Comment by michael-e** [comment]
What's the plan with this issue? (It has not been considered for 2.6)

**Comment by brendo** [comment]
Config option seems like the best bet.

**Comment by michael-e** [comment]
Yes, I think so. So when to include this? 2.6 RC is out, so shall we add this to 3.0?

**Comment by brendo** [comment]
It may be a candidate for a 2.6.1, or yeah 3.0.0.

**Comment by jensscherbl** [comment]
> Config option seems like the best bet.

Seriously?

What about an XSRF extension that provides and validates a token via an event for the frontend? Could use the core's function to generate and validate the token. But with an event, a session would only be generated when the user hits a page where XSRF protection is actually necessary, and not for the entire frontend. Also better for caching.

**Comment by nitriques** [comment]
> What about an XSRF extension that provides and validates a token via an event for the frontend? 

This is the way to go. I've postpone this issue to 2.8.0, but in fact, I agree with @jensscherbl on this. A event is exactly what you need. I even question the whole presence of the `cookie-*` values in the parameters.

**Comment by michael-e** [comment]
I am not sure.

1. We already have an event to validate the token. It's provided by the Cross-Site Scripting (XSS) Filter extension. So this event could be moved to a new extension. But it must be noted that XSRF protection may be circumvented by XSS attacks. [Wikipedia](https://en.wikipedia.org/wiki/Cross-site_request_forgery) says:

    > Cross-site scripting (XSS) vulnerabilities (even in other applications running on the same domain) allow attackers to bypass most CSRF preventions.

    That might give a reason to keep the event in the XSS extension (because this enforces XSS protection if you want XSRF functionality). Choosing a better name for the extension may be a good idea, however.

2. The new extension (or the existing one, see above) would also provide a token on **every page** (unless this can be triggered by a separate event, making things more complicated for the developer). It should do that without duplicating any security-related code from the core. If the token is on every page, the effect will be the same as with my proposal: There's always a session. Providing the token on certain pages only has no big advantage: Once you have seen a token, your session will last (because a non-empty session will never be dropped by Symphony).

    Using XSRF tokens on the frontend will always mean that you need sessions/cookies. And this makes simple frontend caching (using Varnish or nginx) impossible. Well, that's the price! (If you want to protect a web application, your app will need logged-in users with sessions/cookies anyway.)

BTW, I have already implemented XSRF protection in one of my web apps using the XSS Filter extension and the simple Symphony hack proposed above. It works.

> I even question the whole presence of the cookie-* values in the parameters.

I don't really understand what you mean here. An XSRF token is simply tied to a session, and it must be, AFAIK.

**Comment by michael-e** [comment]
While I had lost sight of this, it is still an issue for LTS version users. @calin-borbeli posted on Gitter, and he noticed that this bug is not mentioned in the README.

For 2.7.x, shouldn't we either add s.th. to the README or fix it as discussed (using a config setting)?
